swarm.go

```diff
@@ -98,12 +107,7 @@ type Swarm struct {
 }
 
 // NewSwarm constructs a Swarm.
-//
-// NOTE: go-libp2p will be moving to dependency injection soon. The variadic
-// `extra` interface{} parameter facilitates the future migration. Supported
-// elements are:
-//  - connmgr.ConnectionGater
-func NewSwarm(local peer.ID, peers peerstore.Peerstore, bwc metrics.Reporter, extra ...interface{}) *Swarm {
+func NewSwarm(local peer.ID, peers peerstore.Peerstore, bwc metrics.Reporter, opts ...Option) *Swarm {
```

Comment:
The bandwidth is also optional. We can also make the peerstore optional, but then we'd need to make sure to close it iff we create it.

Reply:
I'd also consider allowing this to return an error.

Reply:
I converted the `metrics.Reporter` into an option. It doesn't look like the peerstore is optional at the moment, so why don't we just keep it as a parameter on the constructor?

Reply:
> so why don't we just keep it as a parameter on the constructor?

No real reason. Making it optional and constructing it internally may make things a bit easier for the user, but it's likely too annoying to be worth it.